Adding (almost) complete Cloud Backup / Cloud Databases

WalterLenuzza opened the pull request:
* Cloud Backup
* Cloud Databases

briancurtin commented on rackspace/backup/v1/_proxy.py:
```diff
@@ -0,0 +1,174 @@
+                           comprised of the properties on the
+                           BackupConfiguration class.
+        :returns: The results of backup_configuration creation
+        :rtype: :class:``~rackspace.backup.v1.backup_configuration
```
extra backtick between :class: and ~rackspace

(I realize docs aren't being built right now and this wouldn't even show up anywhere at the moment, but I'd rather catch it now)

briancurtin commented on rackspace/backup/v1/_proxy.py:
```diff
@@ -0,0 +1,174 @@
+                           comprised of the properties on the
+                           RestoreConfiguration class.
+        :returns: The results of restore_configuration creation
+        :rtype: :class:``~rackspace.backup.v1.restore_configuration
```
another extra backtick

briancurtin commented on rackspace/backup/v1/activity.py:
```diff
@@ -0,0 +1,48 @@
+    #: startRequested, Stoprequested, Completed WithErrors, and Preparing.
+    status = resource.prop('CurrentState')
+    #: Indicates the time of the activity
+    time = resource.prop('TimeOfActivity')
```
In a bunch of other places throughout the upstream SDK, we have timestamps named as things like `updated_at` or `created_at`, but `activity_at` doesn't really fit that. Could this be `timestamp` instead?

briancurtin commented on rackspace/backup/v1/activity.py:
```diff
@@ -0,0 +1,48 @@
+    # capabilities
+    allow_list = True
+
+    # Properties
```
Could probably add `is_backup_configuration_deleted = resource.prop("IsBackupConfigurationDeleted")` in here as well.

briancurtin commented on rackspace/backup/v1/agent.py:
```diff
@@ -0,0 +1,95 @@
+    #: Valid values are IAD, ORD, DFW, HKG, LON, or SYD).
+    datacenter = resource.prop('BackupDatacenter')
+    #: Indicates if the Rackspace Cloud Backup agent on the server is disabled
+    disabled = resource.prop('IsDisabled')
```
We've been doing bool options like `is_disabled`, mainly to avoid conflicts with a few other services that also do timestamps to correspond to the bool, e.g., `disabled_at` for the timestamp and `is_disabled` for the flag.

briancurtin commented on rackspace/backup/v1/agent.py:
```diff
@@ -0,0 +1,95 @@
+    # Properties
+    #: Indicates whether a cleanup can be manually triggered
+    #: on the backup vault
+    cleanup = resource.prop('CleanupAllowed')
```
I feel like this is maybe better named as `can_cleanup` - thoughts?

briancurtin commented on rackspace/backup/v1/agent.py:
```diff
@@ -0,0 +1,95 @@
+    #: Indicates if the Rackspace Cloud Backup agent on the server is disabled
+    disabled = resource.prop('IsDisabled')
+    #: Indicates if backups are encrypted. Valid values are true or false.
+    encrypted = resource.prop('IsEncrypted')
```
`is_encrypted`

briancurtin commented on rackspace/backup/v1/agent.py:
```diff
@@ -0,0 +1,95 @@
+    #: Status of the Cloud Backup agent. Valid values are Online or Offline.
+    status = resource.prop('Status')
+    #: Time of last successful backup
+    time = resource.prop('TimeOfLastSuccessfulBackup')
```
This is one where `last_successful_backup_at` would fit with the usual conventions

briancurtin commented on rackspace/backup/v1/agent.py:
```diff
@@ -0,0 +1,95 @@
+        :type session: :class:`~openstack.session.Session`
+        :returns: ``None``
+        """
+        body = {'MachineAgentId': self.id, "Enable": True}
```
Again not having run this against a live cloud, it looks like the True/False may be flipped here as disable has enable=true and enable has enable=false. Can you confirm this is right?

WalterLenuzza replied:
Yes, you are absolutely correct @briancurtin

briancurtin commented on rackspace/backup/v1/backup_configuration.py:
```diff
@@ -0,0 +1,55 @@
+
+    # Properties
+    #: Indicates if a backup configuration is active.
+    active = resource.prop('IsActive')
```
`is_active`

briancurtin commented on rackspace/backup/v1/backup_configuration.py:
```diff
@@ -0,0 +1,55 @@
+    #: ID that uniquely identifies a Cloud Backup agent
+    agent_id = resource.prop('MachineAgentId')
+    #: Indicates if the backup configuration is deleted
+    deleted = resource.prop('IsDeleted')
```
`is_deleted`

briancurtin commented on rackspace/backup/v1/backup_configuration.py:
```diff
@@ -0,0 +1,55 @@
+    #: Indicates if the backup configuration is deleted
+    deleted = resource.prop('IsDeleted')
+    #: Indicates if backups are encrypted. Valid values are true or false.
+    encrypted = resource.prop('IsEncrypted')
```
`is_encrypted`

briancurtin commented on rackspace/backup/v1/backup_configuration.py:
```diff
@@ -0,0 +1,55 @@
+from rackspace.backup import backup_service
+
+
+class BackupConfiguration(resource.Resource):
```
It looks like there's a bunch of required properties in https://developer.rackspace.com/docs/cloud-backup/v1/developer-guide/#create-backup-configuration that are missing here -- is that a matter of incorrect documentation or something else?

WalterLenuzza replied:
I'm in the process of adding all properties

briancurtin commented on rackspace/backup/v1/restore_configuration.py:
```diff
@@ -0,0 +1,41 @@
+
+    # Properties
+    #: Identifies a unique backup
+    backup = resource.prop('BackupId')
```
`backup_id`

briancurtin commented on rackspace/backup/v1/restore_configuration.py:
```diff
@@ -0,0 +1,41 @@
+    #: Specifies the path where you want the backup to restore
+    destination_path = resource.prop('DestinationPath')
+    #: Indicates if files are overwritten
+    overwrite = resource.prop('OverwriteFiles')
```
I feel like this should be `can_overwrite`

briancurtin commented on rackspace/database/v1/backup.py:
```diff
@@ -0,0 +1,28 @@
+from rackspace.database import database_service
+
+
+class Backup(resource.Resource):
```
This looks like it's missing a few possible properties from https://developer.rackspace.com/docs/cloud-databases/v1/developer-guide/#create-backup

briancurtin commented on rackspace/database/v1/backup_schedule.py:
```diff
@@ -0,0 +1,28 @@
+from rackspace.database import database_service
+
+
+class BackupSchedule(resource.Resource):
```
This also shows a bunch of properties that should be in here https://developer.rackspace.com/docs/cloud-databases/v1/developer-guide/#create-scheduled-backup

briancurtin commented:
Overall this looks ok. There's mostly just some naming issues, which I apologize for nit picking, but I haven't written up those guidelines in our contributor guide yet, but we're trying to make sure everything is as consistent as possible.

One additional set of tests we'll need are unit tests for each of the resources that were added. The proxy unit tests look good, but we need them for the resources as well.

Something I still need to do is figure out how to do some functional testing with a live public cloud, but haven't gotten around to. Do you happen to have any scripts or examples you have run against Rackspace that we can maybe use to prove this works in real life? On the OpenStack SDK this is fairly easy because we startup devstack in a VM, do a bunch of operations on it, then kill the VM. That's a little more difficult in public cloud, which I'm going to be working on next, but really any code you have that runs live would be useful to see so we make sure this works beyond just unit tests.

Thanks a lot for getting this going!

WalterLenuzza commented:
@briancurtin :
Thank for reviewing the above!
I'm not done with fixing Cloud Databases but let know what you think about my latest commit. Should be pretty self-explanatory.

Regarding functional testing: I could help out with a Heat template that stands up most of the needed instances in the public cloud. AutoScale, Load Balancer, Servers (including Backup and Monitoring agents). I suspect tthere are a few products currently not fully supported by Orchestation (HA Cloud Database instances springs to mind). Thoughts?

briancurtin commented:
What we really should do is use the existing functional test ideas from the upstream OpenStack SDK, but what we need is code that runs before and after the tests to ensure that we start and end with a clean environment. In upstream, if a test fails and leaves a server running, it doesn't *really* matter because it's running in a temporary devstack VM as part of our CI system, and that entire thing gets wiped out regardless after the tests run. In our case with this extension, if we leave a server running in public cloud -- and since we serve public cloud we should be running on the real live public cloud -- we're paying for it, and it'll also mess up subsequent runs.

What would probably work best for this is to create a cleanup module per service that just lists every type of resource and then deletes it. That'd run in the setUpClass and tearDownClass methods for each functional test class. I'm going to take a look at this right now with a small example and see what happens.

WalterLenuzza commented:
Sounds good to me. Wiping all resources in an account is straightforward enough. Should we need to keep some resources on the account we can always use metadata.

As part of the [last commit](https://github.com/WalterLenuzza/rackspace-sdk-plugin/commit/f293e8784e18f3d6447f122de9c41eb6c264c508) I've added [unit tests](https://github.com/WalterLenuzza/rackspace-sdk-plugin/tree/master/rackspace/tests/backup/v1) for all of Cloud Backup's resources.
If you could review them and let me know if I'm going down the right path, I'll also add tests for
Cloud Databases.

WalterLenuzza commented:
@briancurtin I've added unit tests for all resources for these two services and improved naming/doc consistency.

briancurtin commented on rackspace/database/v1/ha.py:
```diff
@@ -0,0 +1,189 @@
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
```
This module is better as high_availability_instance.py as a parallel to instance.py.

briancurtin commented on rackspace/database/v1/ha.py:
```diff
@@ -0,0 +1,189 @@
+from rackspace.database import database_service
+
+
+class HA(resource.Resource):
```
Can this be HighAvailabilityInstance?

briancurtin commented:
Overall I think this looks really good save for the two comments I just put in on the HA names being expanded slightly for clarity. I'm going to load it up and run with this and try some things out, and then if the names are updated I think we can accept this.

Additionally, and I think we can maybe solve this in a second PR, is that methods on something like the current HA resource aren't entirely useful only being on the resource itself -- they should be exposed at the Proxy level. If a customer wants to use them, they're going to have to obtain a Session instance in order to call into those, so it's not entirely complete or fully usable to most people. We're not great about having those all fully exposed in the upstream SDK but are working on it, but I just wanted to pass that on here as well.

WalterLenuzza commented:
@briancurtin 
* Added methods to enable/disable Cloud Backup Configurations (+tests) that were missing
* Fixed a few of occurrences where assertions were not entirely useful
* Clarified HA names

If you don't mind reviewing and accepting this, I will make sure all of the resources' methods are exposed on their proxies in a new PR.

briancurtin commented:
One more comment on MagicMocks and then I think we're good. Exposing in a second PR is great.

WalterLenuzza commented:
Replaced all MagicMocks, including [this](https://github.com/rackerlabs/rackspace-sdk-plugin/blob/master/rackspace/tests/object_store/v1/test_bulk_delete.py#L34) and [this](https://github.com/rackerlabs/rackspace-sdk-plugin/blob/master/rackspace/tests/object_store/v1/test_proxy.py#L38).

briancurtin commented:
Alright, let's go forward with this. Once it's through I'll put together a new release and followup via email on a few things.
